pkg/operator/operator.go

```diff
@@ -811,6 +811,33 @@ func (op *Reporting) newPrometheusConn(promConfig promapi.Config) (prom.API, err
 	return prom.NewAPI(client), nil
 }
 
+func (op *Reporting) handleExpiredReport(report *metering.Report, now time.Time) error {
+	// check if the Report is being deleted already
+	if report.DeletionTimestamp != nil {
+		return fmt.Errorf("report was already marked for deletion")
+	}
+	// check if the Report is past its retention time
+	if reportExpired := isReportExpired(op.logger, report, now); !reportExpired {
+		op.logger.Debugf("the %s Report in the %s namespace has not reached the expiration date", report.Name, report.Namespace)
+		return nil
+	}
+	// check if the Report is used by any other Report or ReportQuery, if not delete it
+	if reportIsNotInput := isReportNotUsedAsInput(report, op); reportIsNotInput {
+		newDeletionTimestamp := metav1.Now()
+		report.SetDeletionTimestamp(&newDeletionTimestamp)
+		err := op.meteringClient.MeteringV1().Reports(report.Namespace).
+			Delete(context.TODO(), report.Name, metav1.DeleteOptions{})
+		if err != nil {
+			err = fmt.Errorf("should have deleted report: %s, because expired, but failed to delete", report.Name)
+			return err
+		}
```

Comment:
I realize we may want to amend this error check, and only fire an error when it's not of type `apierrrors.IsNotFound(err)` and handle that kind of error separately.

Reply:
That way, if we're processing a resource that's already been deleted, we can just exit early (or re-queue for later.)

Reply:
cleaned up in 9be2b316